packet/eip_controlplane_reconciliation.go

```diff
@@ -0,0 +1,167 @@
+package packet
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"time"
+
+	"errors"
+
+	"github.com/packethost/packngo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+/*
+ controlPlaneEndpointManager checks the availability of an elastic IP for
+ the control plane and if it exists the reconciliation guarantees that it is
+ attached to a healthy control plane.
+
+ The general steps are:
+ 1. Check if the passed ElasticIP tags returns a valid Elastic IP via Packet API.
+ 2. If there is NOT an ElasticIP with those tags just end the reconciliation
+ 3. If there is an ElasticIP use the kubernetes client-go to check if it
+ returns a valid response
+ 4. If the response returned via client-go is good we do not need to do anything
+ 5. If the response if wrong or it terminated it means that the device behind
+ the ElasticIP is not working correctly and we have to find a new one.
+ 6. Ping the other control plane available in the cluster, if one of them work
+ assign the ElasticIP to that device.
+ 7. If NO Control Planes succeed, the cluster is unhealthy and the
+ reconciliation terminates without changing the current state of the system.
+*/
+type controlPlaneEndpointManager struct {
+	inProcess     bool
+	apiServerPort int
+	eipTag        string
+	instances     cloudInstances
+	deviceIPSrv   packngo.DeviceIPService
+	ipResSvr      packngo.ProjectIPService
+	projectID     string
+	httpClient    *http.Client
+	k8sclient     kubernetes.Interface
+}
+
+func (m *controlPlaneEndpointManager) name() string {
+	return "controlPlaneEndpointManager"
+}
+
+func (m *controlPlaneEndpointManager) init(k8sclient kubernetes.Interface) error {
+	m.k8sclient = k8sclient
+	klog.V(2).Info("controlPlaneEndpointManager.init(): enabling BGP on project")
+	return nil
+}
+
+func (m *controlPlaneEndpointManager) nodeReconciler() nodeReconciler {
+	return m.reconciler()
```

Comment:
This is a bit convoluted. Why not just return it without invoking it?

```go
return m.reconciler
```

And then `reconciler()` can be:

```go
func (m *controlPlaneEndpointManager) reconciler(nodes []*v1.Node, remove bool) error {
    klog.V(2).Info("....")
    if m.inProcess {
// etc.
```

Reply:
`if m.inProcess {` it is not better to return as early as possible from the condition in order to avoid a level of nesting?
```
if !m.inProcess {
    return
}
```

?

Reply:
yes, but I usually like to at least have the log, so when you run with `--v=2` you can know that it got there.

Reply:
The nesting shouldn't matter. This is a once a minute loop, very inexpensive for one more stack push-pop, the optimizer should handle it, and in any case, consistency is better for readability.